src/blog/articles/javascript-async-iterators/index.md

```diff
@@ -0,0 +1,689 @@
+
+Note how this is quite similar to the synchronous version of the iterator protocol. The main difference here is that the `next()` function won't return an object straight away. Instead, it will return a promise that will eventually resolve to an object.
+
+Let's now revisit our countdown example and let's say we want some time to pass before numbers are _produced_:
+
+```javascript
+import { setTimeout } from 'timers/promises'
+
+function createAsyncCountdown (from, delay = 1000) {
+  let nextVal = from
+  return {
+    async next () {
+      await setTimeout(delay)
+      if (nextVal < 0) {
+        return { done: true }
+      }
+
+      return { done: false, value: nextVal-- }
+    }
+  }
+}
+```
+
+Note that this time we are using an _async_ function to implement `next()`. This will make this method immediately return a promise, that will later resolve when we run one of the `return` statements from within the _async_ function.
+
+Also, note that here we are using `setTimeout` from `timers/promises`, a new core module available from Node.js 16.
+
+Ok, now we are ready to use this iterator:
+
+```javascript
+const countdown = createAsyncCountdown(3)
+console.log(await countdown.next()) // { done: false, value: 3 }
+console.log(await countdown.next()) // { done: false, value: 2 }
+console.log(await countdown.next()) // { done: false, value: 1 }
+console.log(await countdown.next()) // { done: false, value: 0 }
+console.log(await countdown.next()) // { done: true }
+```
+
+This works very similarly to its synchronous counterpart with two notable exceptions:
+
+  - We need to use `await` to wait for the next element to be produced.
+  - Between one element and another about 1 second will pass, so this iteration is much slower.
+
+{% image './blog/articles/javascript-async-iterators/javascript-async-iterator-countdown.gif', 'An example of JavaScript async iterator', { maxWidth: 600 }  %}
+
+Of course, here we can use generators as well as a nice syntactic sugar:
+
+```javascript
+import { setTimeout } from 'timers/promises'
+
+async function * createAsyncCountdown (from, delay = 1000) {
+  for (let i = from; i >= 0; i--) {
+    await setTimeout(delay)
+    yield i
+  }
+}
+```
+
+This code is more concise and probably more readable, at least to those accustomed to async functions and generator functions.
+
+
+## The JavaScript async iterable protocol
+
+Let's now discuss the last iteration protocol: the _async iterable protocol_!
+
+> An object is an **async iterable** if it implements the `@@asyncIterator` method, a zero-argument function that returns an **async iterator**.
+
+Note that with `@@asyncIterator` we indicate a symbol that can be accessed with the global value `Symbol.asyncIterator`.
+
+Once again, this definition is quite similar to its synchronous counterpart. The main difference is that this type we have to use `Symbol.asyncIterator` and that it must return an _async_ iterator.
```

Comment:
"this type" => "this time"
"that it must return..." => "to define a function that returns and async iterator"